app/import_functions.go

```diff
@@ -1118,11 +1119,27 @@ func (a *App) importReplies(c *request.Context, data []ReplyImportData, post *mo
 }
 
 func (a *App) importAttachment(c *request.Context, data *AttachmentImportData, post *model.Post, teamID string) (*model.FileInfo, *model.AppError) {
-	file, err := os.Open(*data.Path)
-	if file == nil || err != nil {
-		return nil, model.NewAppError("BulkImport", "app.import.attachment.bad_file.error", map[string]interface{}{"FilePath": *data.Path}, "", http.StatusBadRequest)
+	var (
+		name string
+		file io.Reader
+	)
+	if data.Data != nil {
+		zipFile, err := data.Data.Open()
+		if zipFile == nil || err != nil {
+			return nil, model.NewAppError("BulkImport", "app.import.attachment.bad_file.error", map[string]interface{}{"FilePath": *data.Path}, "", http.StatusBadRequest)
+		}
+		defer zipFile.Close()
+		name = data.Data.Name
+		file = zipFile.(io.Reader)
+	} else {
+		realFile, err := os.Open(*data.Path)
+		if realFile == nil || err != nil {
+			return nil, model.NewAppError("BulkImport", "app.import.attachment.bad_file.error", map[string]interface{}{"FilePath": *data.Path}, "", http.StatusBadRequest)
+		}
+		defer realFile.Close()
+		name = realFile.Name()
+		file = realFile
 	}
-	defer file.Close()
 
 	timestamp := utils.TimeFromMillis(post.CreateAt)
 
```

Comment:
I will look at this more deeply and see how involved that change would be, and if it's simple I will just go ahead and do it, because it _will_ make this change more robust if that's in place. If it looks like it's going to take me any significant amount of time, I'll open a new ticket to track it as follow-up work.

Reply:
>The sha1 can be calculated by copying the file in chunks

Hmm, @streamer45 did you have an implementation in mind for doing this? The only way that I can see to do this would be to write my own implementation of SHA-1 based on the one in the standard library, but that seems pretty involved, and would definitely have to be its own ticket.

Reply:
I was thinking of creating a `hash.Hash` (it's a writer) and compute the sum by direct copy from the file.  
This plus some work on the file uploading part would eliminate the need for the `ioutil.ReadAll()` call.

Something like:

```go
h := sha1.New()
io.Copy(h, file)
sum := h.Sum(nil)
```

Reply:
Ah, I did not see that the `hash.Hash` type was a `Writer`. I thought I might be overlooking something along those lines. I will take another look.

Reply:
>The sha1 can be calculated by copying the file in chunks and we have means to upload a file by streaming it as well. I think it's worth opening a ticket.

Ah okay, I've had a chance to sink my teeth into this a bit more and I see what you mean about "we have means to upload the file by streaming it as well," however doing so does seem to me, at the very least, to be an independent chunk of work from this change, because of the modifications that need to be made to `DoUploadFile` and the methods on which it depends. 

I think trying to add this functionality to this PR would just make this PR confusing, so I've gone ahead and opened [a ticket that you can find here](https://mattermost.atlassian.net/browse/MM-36241) to track this follow-up work. If this becomes a high priority, feel free to assign that ticket to me.

Reply:
I think it makes sense to leave this to a separate ticket. But just for some added context I think this can be done mainly in two ways:

One is to use `app.UploadFileX()`, essentially replicating what `api4.uploadFileSimple()` does. It streams the file directly from the request to the backing store.

Otherwise we can use the newer upload API. It's a bit more verbose but supports resuming uploads. You first create an upload session through `app.CreateUploadSession()` and then call `app.UploadData()` which accepts a reader and again it's all streamed.